### DISCUSSION TITLE
Using models to type dict keys

### DISCUSSION THREAD

**Discussion by awgymer** [pydantic/pydantic#10742] [Question]
I have the following model:
```py
from pydantic import BaseModel

InfoVal = str | int | float | bool
InfoTuple = tuple[str, ...] | tuple[int, ...] | tuple[float, ...] | tuple[bool, ...]
InfoDict = dict[str, InfoVal | InfoTuple]

class ModelWithInfo(BaseModel):
    id: str
    qual: float | None
    val: str
    info: InfoDict
```

This works fine. But now I would like to be able to type-check the `info` field for certain key/value pairs, whilst allowing extra arbitrary keys (so a `TypedDict` will not work). 

I envision something like:
```py
from pydantic import ConfigDict

class ModelInfo(BaseModel):
    model_config = ConfigDict(extra="allow")

    fieldA: str
    fieldB: int
    fieldC: tuple[float, float]
    fieldD: bool
    
class ModelWithInfo(BaseModel):
    id: str
    qual: float | None
    val: str
    info: ModelInfo
```

This would work, however I would like to allow arbitrary subclassing of `ModelWithInfo` such that any `ModelInfo` class that has attributes only taking `InfoVal | InfoTuple` would be able to be supplied to info:

```py
class ModelWithInfo(BaseModel):
    id: str
    qual: float | None
    val: str
    info: ModelInfoAbstractBase
 
class ModelInfoA(BaseModel):
    model_config = ConfigDict(extra="allow")
    a: str

class ModelWithInfoA(ModelWithInfo):
    info: ModelInfoA
```

I cannot fathom how such a design can be achieved with pydantic though?

**Comment by Viicos** [comment]
Not sure I understand your use case, but you can make use of generics:

```python
ModelInfoT = TypeVar('ModelInfoT', bound=ModelInfo)

class ModelWithInfo(BaseModel, Generic[ModelInfoT]):
    id: str
    qual: float | None
    val: str
    info: ModelInfoT

ModelWithInfoA = ModelWithInfo[ModelInfoA]
```

**Reply by awgymer** [reply]
But how to restrict `ModelInfo` to be effectively a generic basemodel over multiple types. 
Something like the following (it doesn't work I tried):
```py
Modelnfo = BaseModel[InfoVal | InfoTuple]

class ModelWithInfo(BaseModel, Generic[ModelInfo]):
    id: str
    qual: float | None
    val: str
    info: ModelInfo
```

`ModelInfo` should be able to have any attribute names, but the value types need to be restricted. Effectively a `TypedDict` that allows extra arbitrary keys.

**Reply by Viicos** [reply]
What about https://docs.pydantic.dev/dev/concepts/models/#extra-fields?

**Reply by awgymer** [reply]
That alone does not solve the issue. Which is how to type a model generically without defining a set of required attribute names. 

It's some combination of `dict[str, InfoVal | InfoTuple]` and either `BaseModel` (which allows extra fields) or `TypedDict` (which is less likely to work because you can't allow extra fields)

**Reply by Viicos** [reply]
Honestly it's still unclear what's your use case here, but you might be interested in https://peps.python.org/pep-0728/, still in draft.

**Reply by awgymer** [reply]
I will try to give a better example using `BaseModel`:
```py
DictVal = str | int | float
CustomDict = dict[str, DictVal]

class ModelWithInfo(BaseModel):
    id: str
    val: int
    info: <generic type hint that allows any BaseModel subclass where the defined attrs are only of DictVal types>

class InfoA(BaseModel):
    a: int
    b: str

class InfoB(BaseModel):
    c: bool
    d: dict[int, str]

class ModelInfoA(ModelWithInfo):
    info: InfoA # THIS WOULD TYPE CORRECTLY AND WORK WITH EG MYPY/PYRIGHT
    
class ModelInfoB(ModelWithInfo):
   info: InfoB # THIS WOULD BE FLAGGED AS A TYPE ERROR BECAUSE InfoB.d is NOT A DICTVAL TYPE
```


Extra items in a `TypedDict` might be a potential aid in this scenario but you would still need be able to type hint e.g. `TypedDict[str, DictVal]` which does not work. 

The reason `info` cannot be a plain `CustomDict` type hint is that I want to be able to enforce specific keys (and value types) for subclasses (whilst allowing additional items)

You could just define each model without a shared base, but in reality the base model does not just have some attributes to be shared but also some helper class methods (which would need to know about `info` type and thus that field can't be left off of the basemodel but needs to be generically/abstractly typed to allow any allowable subclass).

**Reply by awgymer** [reply]
Ok I think that `Generic` gets us most of the way:
```py
class ModelInfo(BaseModel):
    __pydantic_extra__: InfoDict | None = Field(init=False)
    model_config = ConfigDict(extra="allow")


class ModelInfoA(ModelInfo):
    model_config = ConfigDict(extra="forbid")

    x: int
    b: str


ModelInfoT = TypeVar("ModelInfoT", bound=ModelInfo)


class ModelWithInfo(BaseModel, Generic[ModelInfoT]):
    id: str
    qual: float | None
    val: str
    info: ModelInfoT


ModelWithInfoA = ModelWithInfo[ModelInfoA]
```

Sadly this types correctly with `pyright` and `mypy` but I would like to force the `ModelInfoT` bound type to enforce `extra="allow"` and also as above, to enforce that it only ever has attributes with types of `str | int | float | bool` - because my functions which take the `ModelInfoT` objects would expect `model_dump` to conform to `dict[str, str | int | float | bool]`

Does this make more sense at all?
